crates/host/src/env.rs

```diff
@@ -97,4 +101,84 @@ impl Env {
             }
         }
     }
+
+    /// Mimics upstream function of the same name but accesses the global directly from env.
+    /// https://github.com/wasmerio/wasmer/blob/master/lib/middlewares/src/metering.rs#L285
+    pub fn get_remaining_points(
+        &self,
+        store_mut: &mut StoreMut,
+    ) -> Result<MeteringPoints, wasmer::RuntimeError> {
+        let exhaused: i32 = self
+            .wasmer_metering_points_exhausted
+            .as_ref()
+            .ok_or(wasm_error!(WasmErrorInner::Memory))?
+            .get(store_mut)
+            .try_into()
+            .map_err(|_| wasm_error!(WasmErrorInner::PointerMap))?;
+
+        if exhaused > 0 {
+            return Ok(MeteringPoints::Exhausted);
+        }
+
+        let points = self
+            .wasmer_metering_remaining_points
+            .as_ref()
+            .ok_or(wasm_error!(WasmErrorInner::Memory))?
+            .get(store_mut)
+            .try_into()
+            .map_err(|_| wasm_error!(WasmErrorInner::PointerMap))?;
+
+        Ok(MeteringPoints::Remaining(points))
+    }
+
+    pub fn set_remaining_points(
+        &self,
+        store_mut: &mut StoreMut,
+        points: u64,
+    ) -> Result<(), wasmer::RuntimeError> {
+        self.wasmer_metering_remaining_points
+            .as_ref()
+            .ok_or(wasm_error!(WasmErrorInner::Memory))?
+            .set(store_mut, points.into())
+            .map_err(|_| wasm_error!(WasmErrorInner::PointerMap))?;
+
+        self.wasmer_metering_points_exhausted
+            .as_ref()
+            .ok_or(wasm_error!(WasmErrorInner::Memory))?
+            .set(store_mut, 0i32.into())
+            .map_err(|_| wasm_error!(WasmErrorInner::PointerMap))?;
+        Ok(())
+    }
+
+    pub fn decrease_points(
+        &self,
+        store_mut: &mut StoreMut,
+        points: u64,
+    ) -> Result<MeteringPoints, wasmer::RuntimeError> {
+        match self.get_remaining_points(store_mut) {
+            Ok(MeteringPoints::Remaining(remaining)) => {
+                if remaining < points {
+                    self.wasmer_metering_remaining_points
+                        .as_ref()
+                        .ok_or(wasm_error!(WasmErrorInner::Memory))?
+                        .set(store_mut, 0i32.into())
```

Comment:
It looks like that's not the idea. You allocate a certain amount of capacity and when it's gone it's gone until something calls `set` to top it back up?

I still wonder if accurately tracking the value to account for overrun and making the 'set' an 'increment by' makes sense?

Reply:
@ThetaSinner well changing the middleware from decreasing points to increasing would mean rewriting the middleware from the ground up

we're using the default setup from wasmer, and what i'm doing here is exposing it so that it can be accessed by host functions as well as the default wasm op counter

Reply:
Sorry, I don't mean dropping the increasing part, the increasing part is fine. I'm looking at `set_remaining_points` and wondering if that should be a paired up increment rather than a set?

If this is just exposing something wasmer already does then fine

Reply:
@ThetaSinner it's a direct port of https://github.com/wasmerio/wasmer/blob/master/lib/middlewares/src/metering.rs#L335 but not requiring the instance itself to be in the environment, all we need is the global

Reply:
Understood thanks!